common/interpolation/src/spline_interpolation.cpp

```diff
@@ -166,53 +140,59 @@ std::vector<double> splineByAkima(
 }
 }  // namespace interpolation
 
+Eigen::Index SplineInterpolation::get_index(double key) const
+{
+  auto it = std::lower_bound(base_keys_.begin(), base_keys_.end(), key);
+  return std::clamp(
+    static_cast<int>(std::distance(base_keys_.begin(), it)) - 1, 0,
+    static_cast<int>(base_keys_.size()) - 2);
+}
+
 void SplineInterpolation::calcSplineCoefficients(
   const std::vector<double> & base_keys, const std::vector<double> & base_values)
 {
   // throw exceptions for invalid arguments
   interpolation_utils::validateKeysAndValues(base_keys, base_values);
-
-  const size_t num_base = base_keys.size();  // N+1
-
-  std::vector<double> diff_keys;    // N
-  std::vector<double> diff_values;  // N
-  for (size_t i = 0; i < num_base - 1; ++i) {
-    diff_keys.push_back(base_keys.at(i + 1) - base_keys.at(i));
-    diff_values.push_back(base_values.at(i + 1) - base_values.at(i));
-  }
-
-  std::vector<double> v = {0.0};
-  if (num_base > 2) {
-    // solve tridiagonal matrix algorithm
-    TDMACoef tdma_coef(num_base - 2);  // N-1
-
-    for (size_t i = 0; i < num_base - 2; ++i) {
-      tdma_coef.b[i] = 2 * (diff_keys[i] + diff_keys[i + 1]);
-      if (i != num_base - 3) {
-        tdma_coef.a[i] = diff_keys[i + 1];
-        tdma_coef.c[i] = diff_keys[i + 1];
-      }
-      tdma_coef.d[i] =
-        6.0 * (diff_values[i + 1] / diff_keys[i + 1] - diff_values[i] / diff_keys[i]);
-    }
-
-    const std::vector<double> tdma_res = solveTridiagonalMatrixAlgorithm(tdma_coef);
-
-    // calculate v
-    v.insert(v.end(), tdma_res.begin(), tdma_res.end());
-  }
-  v.push_back(0.0);
-
-  // calculate a, b, c, d of spline coefficients
-  multi_spline_coef_ = interpolation::MultiSplineCoef{num_base - 1};  // N
-  for (size_t i = 0; i < num_base - 1; ++i) {
-    multi_spline_coef_.a[i] = (v[i + 1] - v[i]) / 6.0 / diff_keys[i];
-    multi_spline_coef_.b[i] = v[i] / 2.0;
-    multi_spline_coef_.c[i] =
-      diff_values[i] / diff_keys[i] - diff_keys[i] * (2 * v[i] + v[i + 1]) / 6.0;
-    multi_spline_coef_.d[i] = base_values[i];
+  Eigen::VectorXd x = Eigen::Map<const Eigen::VectorXd>(
+    base_keys.data(), static_cast<Eigen::Index>(base_keys.size()));
+  Eigen::VectorXd y = Eigen::Map<const Eigen::VectorXd>(
+    base_values.data(), static_cast<Eigen::Index>(base_values.size()));
+
+  const auto n = x.size();
+
+  if (n == 2) {
+    a_ = Eigen::VectorXd::Zero(1);
+    b_ = Eigen::VectorXd::Zero(1);
+    c_ = Eigen::VectorXd::Zero(1);
+    d_ = Eigen::VectorXd::Zero(1);
+    c_[0] = (y[1] - y[0]) / (x[1] - x[0]);
+    d_[0] = y[0];
+    base_keys_ = base_keys;
+    return;
   }
 
+  // Create Tridiagonal matrix
+  Eigen::VectorXd v(n);
+  Eigen::VectorXd h = x.segment(1, n - 1) - x.segment(0, n - 1);
```

Comment:
const

Reply:
Resolved in https://github.com/autowarefoundation/autoware.universe/pull/8903/commits/55abe390ab4c5b34cb65b2baae199b7715d32a74